### PR TITLE
feat: Add integration export and import endpoints.

### LINKED ISSUE #764: Add integration export and import endpoints.
Need new endpoints:
* `GET /api/v1/integrations/{id}/export.zip` should download a .zip file which holds all the data need to re-import the given integration into a different environment.

* `POST /api/v1/integration-support/import` and posting it the exported zip should create/update the integration in the new env.

### PR DESCRIPTION
**Pull request by chirino**
Fixes #764

### CODE REVIEW

**Comment by KurtStam** [comment]
Don't we need a check on the version of the model/api?

**Comment by chirino** [comment]
@KurtStam I think that's something that we can add later. We generally need something like that for all our entities and perhaps we can leverage that here too.

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java [hunk 1]
```diff
@@ -46,4 +63,83 @@ public IntegrationSupportHandler( ProjectGenerator projectConverter) {
...
+            while (true) {
+                ZipEntry entry = zis.getNextEntry();
+                if( entry == null ) {
+                    break;
+                }
+                if (EXPORT_MODEL_FILE_NAME.equals(entry.getName())) {
+                    ModelData[] models = Json.mapper().readValue(new FilterInputStream(zis) {
+                        @Override
+                        public void close() throws IOException {
+                            // We want to avoid closing zis
```
[the remark this answers is not in the document]
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: we don't want to close the zip stream since we wont be able to read from it again when we get to the next zip entry.

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java [hunk 2]
```diff
@@ -46,4 +63,83 @@ public IntegrationSupportHandler( ProjectGenerator projectConverter) {
...
+                        }
+                    }, ModelData[].class);
+
+                    imported += importModels(models);
+                }
+                zis.closeEntry();
+            }
+            if (imported==0) {
+                LOG.info("Could not import integration: No integration data model found.");
+                return Response.status(Response.Status.BAD_REQUEST).build();
```
[the remark this answers is not in the document]
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: We need to think about what kind of error messages would actually help a citizen integrator before we go to deep down this route.

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java [hunk 3]
```diff
@@ -46,4 +63,86 @@ public IntegrationSupportHandler( ProjectGenerator projectConverter) {
...
+                LOG.info("Could not import integration: " + e, e);
+            }
+            return Response.status(Response.Status.BAD_REQUEST).build();
+        }
+    }
+
+    public int importModels(ModelData... models) throws IOException {
+        int count = 0;
+        for (ModelData model : models) {
+            switch (model.getKind()) {
```
Comment (zregvart) [inline rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: Could this be simpler? We can get `modelClass` from `Kind` and could have a constant `EnumSet<Kind> ALLOWED_IMPORTS`?

Comment (zregvart) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: Aah I see Connections/Connectors are handled differently

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 4]
```diff
@@ -108,6 +121,44 @@ public Integration get(String id) {
...
+                Connector connector = getDataManager().fetch(Connector.class, connection.getConnectorId().get());
+                if( connector != null ) {
+                    models.add(new ModelData(Kind.Connector, connector));
+                }
+            }
+        }
+
+        return out -> {
+            try (ZipOutputStream tos = new ZipOutputStream(out) ) {
+                addEntry(tos, EXPORT_MODEL_FILE_NAME, Json.mapper().writeValueAsBytes(models));
```
Comment (zregvart) [inline rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: Would a ZIP file with multiple entries make more sense? the file names could be sanitized names of Integrations/Connectors/Connections?

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 5]
```diff
@@ -108,6 +121,44 @@ public Integration get(String id) {
...
+        Integration integration = this.get(id);
+        models.add(new ModelData(Kind.Integration, integration));
+
+        for (Step step : integration.getSteps()) {
+            Optional<Connection> c = step.getConnection();
+            if( c.isPresent() ) {
+                Connection connection = c.get();
+                models.add(new ModelData(Kind.Connection, connection));
+                Connector connector = getDataManager().fetch(Connector.class, connection.getConnectorId().get());
+                if( connector != null ) {
```
Comment (rhuss) [inline rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: In which case can the connector be null ? Why it is an optional on connection anyway ? (so you should check before calling get())

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 6]
```diff
@@ -108,6 +121,44 @@ public Integration get(String id) {
...
+                    models.add(new ModelData(Kind.Connector, connector));
```
Comment (rhuss) [inline rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: Shouldn't you add a connector only if not already adedd ? (e.g. when you use the same connector but different actions for incoming and outcoming)

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java (hunk 4 above)
Comment (rhuss) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: For the moment I don't expect many people looking into the zip so its somewhat opaque. 

What I miss though is some sort of versioning, which allows for evolving the export/import format over time.

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 7]
```diff
@@ -108,6 +121,44 @@ public Integration get(String id) {
...
+        ArrayList<ModelData> models = new ArrayList<>();
+
+        Integration integration = this.get(id);
+        models.add(new ModelData(Kind.Integration, integration));
+
+        for (Step step : integration.getSteps()) {
+            Optional<Connection> c = step.getConnection();
+            if( c.isPresent() ) {
+                Connection connection = c.get();
+                models.add(new ModelData(Kind.Connection, connection));
```
Comment (rhuss) [inline rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: Shouldnt a connection be added only if not already added ?

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java [hunk 8]
```diff
@@ -46,4 +63,86 @@ public IntegrationSupportHandler( ProjectGenerator projectConverter) {
...
+                        dataManager.update(integration);
+                    }
+                    count ++;
+                    break;
+
+                case Connection:
+
+                    // We only create connections, never update.
+                    Connection connection = (Connection) model.getData();
+                    if (dataManager.fetch(Connection.class, connection.getId().get()) == null) {
```
Comment (rhuss) [inline rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: Is the id() some kind of symbolic name or a real (universal ?) unique id ? How it is is guaranteed that there is no clash with an already existing id (not coming from an import) ? Would it make sense to add some kind of unique prefix / namespace to imported connections to indicate their source ?

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java [hunk 9]
```diff
@@ -108,6 +121,44 @@ public Integration get(String id) {
         return integration;
     }
 
+
+    @GET
+    @Path("/{id}/export.zip")
```
Comment (rhuss) [inline rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: as the name used here is also used for storing the file, I would chose a more expressive name like 'ignite-integration-export.zip' (or 'syndesis-integraiton-export.zip').

Or, alternatively and probably better, set the Http-Header for `Content-Type: application/octet-stream` and `Content-Disposition: attachment; filename="syndesis-integration-export.zip"` (with customization for fuse-ignite in production)

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java (hunk 4 above)
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: I agree, but I think we also need this in the DB model for the same reasons.  So I think once we put that in place there, this export/import thing will likely pick it up for free.

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java (hunk 9 above)
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: It's not really used to control the file name.  The ui module is in control of the the file name.  see: https://github.com/syndesisio/syndesis-ui/blob/master/src/app/integrations/detail-page/detail.component.ts#L413

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java (hunk 7 above)
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: I considered this, but there is no harm in exporting it multiple times and it should be a very unlikely scenario.  Keeping code simple wins in this case.

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java (hunk 5 above)
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: It would be an odd case where the DB is not consistent.  Generally I think all this code could use more null/optional checking.  But this is more of a POC anyways.  (@dsimansk: would like these edge cases to be found by tests)

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java (hunk 6 above)
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationHandler.java]: Once again there is no harm in exporting the same thing multiple times.  It only gets imported once.

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java (hunk 3 above)
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: Right!

File: rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java (hunk 8 above)
Comment (chirino) [reply rest/src/main/java/io/syndesis/rest/v1/handler/integration/IntegrationSupportHandler.java]: Our id's are GUIDs, there should be no clashes so I think this is safe.
matching on the name of the object would be the other option, but is less safe since you could have conflicts, but might make more sense since folks are in control of that more.

**Comment by rhuss** [comment]
I'm wrong WRT the connector ids, as they are already name-like (not a random id as for the integration).

**Comment by rhuss** [comment]
@chirino There's a conflict we you need to resolve.

**Comment by chirino** [comment]
rebased and squashed.
